Predict 2D labels

aazuspan opened the pull request:
This would close #5 by updating the sampling and training workflows to use 2D label data instead of single values. This adds flexibility by allowing us to work with arbitrary NAIP image sizes instead of being tied to a single 30m pixel. To accomplish this, there were some major changes made:

- Switch from sampling local LiDAR files to sampling LiDAR images from Earth Engine. This was done primarily to ensure that we could reliably align footprints to the NAIP data in Earth Engine, but also offloads the extraction work to the cloud. A script was added to pre-process the single-band LiDAR rasters into multi-band rasters which can be uploaded to image collections in Earth Engine. That ingestion could be automated with a tool like [geeup](https://github.com/samapriya/geeup), but I figured it should be a relatively rare task, so it is currently just done manually.
- A new `Acquisition` class was added to handle the loading of LiDAR and NAIP images from Earth Engine. This also utilizes LCMS annual forest loss maps to mask any loss that occurred between LiDAR and NAIP acquisitions.
- The `sampling` module was updated to sample from Earth Engine `Point` geometries instead of x, y coordinates.
- For now, the sampling notebook has been modified to extract *all* undisturbed pixels at the specified minimum spacing over the acquisition footprint, rather than a random sample. These can be subsampled later if desired.
- Previously, sampled values were extracted on-the-fly using parallel `getInfo` calls. Now that we're extracting exponentially more data at each sample, this isn't possible as even small requests lead to memory errors. Instead, we now have to export sampled results to Drive via CSV and reformat the CSVs into HDFs.
- The `Dataset` class (now `NAIPDataset`) was refactored to work with 2D label data. It was also simplified such that the acquisition year is now part of the identifying dataset name, and the number of samples is no longer an identifier because it isn't a user-defined parameter when building the dataset.

### Preliminary results

This was far from a controlled experiment, but results look promising. The table below compares canopy cover results from the baseline `main` branch model and dataset, and with the addition of the new model and new data. *Note this comparison was originally run between 2010 and 2016[^2] data, but has been updated to use the MAL2010 acquisition for everything*.

Model | Dataset | n | MSE | R^2 | Notes
-- | -- | -- | -- | -- | --
CNN_v1 | 1D - MAL2010 | 30k | 201 | 0.55 | Baseline
UNet | 1D - MAL2010 | 30k | 195 | 0.57 | New model, old data
CNN_v1 | 2D - MAL2010 | 450k | 223 | 0.54 | New data, old model
UNet | 2D - MAL2010 subsample[^1] | 30k | 194 | 0.58 | New model and data (same data quantity)
UNet | 2D - MAL2010 | 450k | 161 | 0.65 | **New model and data (proposed)**

The takeaway seems to be that the new model and dataset alone didn't improve performance, but the combination of the two did. The new dataset improved performance after accounting for the larger sample size[^1], suggesting that the increased dimensionality contributed to the improvement.

[^1]: The updated 2D dataset contains only 18,000 samples, but each sample contains 25 measurements, giving us ~15x as much data as the old dataset. The subsampled dataset contains only 1,200 samples with 25 measurements each, giving us 30,000 total LiDAR measurements.
[^2]: With the MAL2016 dataset and the proposed model and data structure, we got MSE 139 and R^2 0.78, suggesting that performance varies quite a bit between datasets.

aazuspan commented:
@grovduck If you have some time and want to try running through all the updated notebooks, that would be great, but if not, no worries! Fair warning that with the updated 150x150m image sizes, the exports take much longer and produce much larger files (the 2016 training CSV used here is ~7GB).

It will take some more experimentation to decide if the larger label size has a benefit outside of the increased amount of training data, but I think that regardless we'll want to use this updated workflow with 2D label data, whether we go back to just 1x1 labels at 30m or go all the way to 150x150 labels with 1m LiDAR data.

grovduck commented:
@aazuspan, I think I have a silly first question.  When running `01_sampling.ipynb`, I was expecting that I wouldn't need to authenticate because of the fix you made in #10.  But, is this code *forcing* the reauthentication?
```
import ee
ee.Authenticate()
ee.Initialize()
```
Should `ee.Authenticate` pass silently if the local credentials are already in place?  The rest of the notebook was a dream :smile:

aazuspan commented:
Good question! `ee.Authenticate` does force re-authentication even if you already have credentials. 

Those calls are mostly just left over from before #10, but I was also a little hesitant to remove them in case a user tried to run the notebooks without local credentials. Now that I think about it though, I'm leaning towards removing them all and just making a note in the first notebook to run `ee.Authenticate` if necessary. What do you think? In the meantime I've just been commenting them out whenever I run those cells, which definitely *isn't* the right solution 😅

grovduck commented:
> In the meantime I've just been commenting them out whenever I run those cells.

Heehee, me too!  

I like your proposed solution.  Get authentication out of the way at the beginning (assuming that a user will run the notebooks in order) and then just have `ee.Initialize()` calls.  I guess the other option is a helper function used everywhere that tests for initialization and then authenticates if an exception is thrown?  But not sure it's worth the effort especially if most users will be authenticated already.

grovduck commented on notebooks/04_inference.ipynb:
```diff
@@ -196,22 +205,24 @@
     ")\n",
     "\n",
-    "with rasterio.open(\"../data/pred/malheur_cnnv1_cancov_.tif\", \"w\", **profile) as dst:\n",
+    "with rasterio.open(f\"../data/pred/malheur_2016_{MODEL_ID}.tif\", \"w\", **profile) as dst:\n",
```
Probably `MODEL_ID` here is sufficient.  In your examples, `MODEL_ID` already starts with `MAL2016`

aazuspan replied:
I agree that the current solution isn't great, but I'm a little hesitant to just go with `MODEL_ID` since the `MAL2016` model could potentially be used to generate a prediction for a different area or year. It seems to me like it should ultimately be identified by both the `MODEL_ID` and an ID associated with the NAIP data used for mapping, but I'm not quite sure how best to handle the latter. Let me know if you have any thoughts!

grovduck replied:
Ahh, this was me not understanding that there was a difference between the first two elements (prediction area + year) and the model used for prediction.  Because this is in a notebook and gives the *example* flow, I'm not sure it's worth formalizing here.  As you say, coming up with an ID to uniquely identify the NAIP data would be useful, but not probably terribly important at this point. 

I confess to grabbing the NAIP 2016 window that covers the lidar acquisition just to see how it behaved where it was trained and didn't put any effort into differentiating the output prediction name.  Just appended a '2' on the end!

aazuspan replied:
Understandable - that distinction is a little ambiguous at the moment!

> I confess to grabbing the NAIP 2016 window that covers the lidar acquisition just to see how it behaved where it was trained

Interesting, how did it do? I've only compared the 2016 prediction to the 2010 and 2020 LiDAR datasets, which obviously aren't the best comparisons due to the temporal mismatch.

grovduck replied:
> Interesting, how did it do? I've only compared the 2016 prediction to the 2010 and 2020 LiDAR datasets, which obviously aren't the best comparisons due to the temporal mismatch.

Patterns look really good to me in terms of capturing fine scale landscape features.  There looks to be an attenuation effect (i.e. regression to the mean) where it's not capturing the highest highs / lowest lows, but that's definitely not uncommon for a lot of modeling.

| Lidar canopy cover | Predicted NAIP canopy cover |
| ----------- | ----------- |
| <img src="https://github.com/lemma-osu/naip-cnn/assets/1706997/1276313c-9c86-4309-bb69-6f88dfabf875" width="400" height="400"> | <img src="https://github.com/lemma-osu/naip-cnn/assets/1706997/be5c34d4-7659-449a-998d-3d4381a499df" width="400" height="400">

aazuspan replied:
Glad to hear! I'm sure there's still room for improvement in the model and training parameters, so hopefully we can match the distributions a little better before we're done.

aazuspan commented:
The helper function is a cool idea, but I agree probably not worth the effort here. I removed the authenticate calls from all notebooks and added a note to the sampling notebook. I also think that trying to initialize without authenticating will give a helpful error, so it should be easy enough to get running in the off chance someone uses this who isn't already authenticated.

grovduck commented on src/naip_cnn/acquisitions.py:
```diff
@@ -0,0 +1,84 @@
+        )
+
+
+# OR NAIP years: 2004 (RGB), 2005 (RGB), 2009, 2011, 2012, 2014, 2016, 2020, 2022
```
More a question than anything.  I'm wondering if you had a rule for associating lidar acquisitions with NAIP acquisitions in terms of date.  For example, the 2007 lidar uses the 2009 NAIP (closest with 4-band data).  It looks like most of the time, you tried to use a later NAIP acquisition, but only if it was closer temporally (e.g. the 2017 lidar acquisitions use 2016 NAIP).  With the LCMS masking in there, this is definitely less of an issue.

aazuspan replied:
Good question, as there's definitely different ways to approach this! As you guessed, I prioritized the closest acquisition in time, and went with the more recent acquisition in the case of a tie. For 2007 that was done to avoid the RGB-only imagery from 2005, but generally I figured that more recent imagery should *theoretically* be higher quality. As you mentioned, masking disturbances should hopefully remove any changes that occur within a couple years, but maybe we should ultimately limit our validation to higher-confidence acquisitions that occurred in the same year?

Another idea I haven't implemented would be to train with multiple NAIP years for each LiDAR acquisition, e.g. a single canopy cover chip from 2018 could be compared against NAIP imagery from both 2016 and 2020, effectively doubling our training data. Given the amount of training data we already have to work with, I'm not sure this is really worth pursuing, but it's something to consider. We would need to be careful not to train with one year and validate with another over the same plot, though...

grovduck replied:
> Another idea I haven't implemented would be to train with multiple NAIP years for each LiDAR acquisition, e.g. a single canopy cover chip from 2018 could be compared against NAIP imagery from both 2016 and 2020, effectively doubling our training data. Given the amount of training data we already have to work with, I'm not sure this is really worth pursuing, but it's something to consider.

I completely defer to you on this, but this seems like it would add some complexities for not a lot of gain.  Given that you're starting to think about how to minimize differences in NAIP collections (#13, #14), that might even serve to produce more-or-less duplicate training data.  But there might be an upside that I haven't considered?

To be clear, I think your current approach is great and preferentially choosing a later NAIP acquisition makes sense to me.

aazuspan replied:
I suspect that even with those normalization steps, there will still be enough difference between a footprint in subsequent NAIP acquisitions (shadow angle, radial distortion, etc.) that it might help the model to generalize further, but I agree that we should probably avoid adding that complexity for now. We've got enough things to worry about already!

grovduck commented on src/naip_cnn/sampling.py:
```diff
@@ -1,227 +1,89 @@
     )
 
-    return pd.concat(gdfs).drop(columns="geometry").reset_index(drop=True)
+    if drop_if_null:
```
I *think* I understand how `drop_if_null` cooperates with the `ee.FeatureCollection.map` parameter `opt_dropNulls`, but probably good for me to talk it through.  If you pass `drop_if_null=True`, any footprint with masked pixels will get a value of None (null) and then `opt_dropNulls` will remove it from the feature collection.  But if you pass `drop_if_null=False`, these partially masked footprints will stay as features.  Will this cause issues when training?

aazuspan replied:
Your understanding is spot on. I'm pretty sure that the model would train just fine with nulls, but I'm not sure about the performance implications. Since we have no shortage of training data and nulls only tend to occur at the edges of the LiDAR data, I figured it's safer just to drop them

grovduck replied:
I was actually thinking that more nulls might get caused by the LCMS mask, but perhaps that's not a huge amount considering the short time frame between NAIP and lidar acquisitions.  But, as you say, you're definitely not lacking for data.

aazuspan replied:
I wasn't thinking about that, but it is of course another source of null values! I don't have a great idea of how many footprints we're losing to that, but it would be good to know - I'll give that a quick look. Another thing to consider is that LCMS likely misses disturbances at the sub-30m pixel scale, so avoiding footprints that are immediately adjacent to masked pixels is probably a safer tactic, as long as we're not limited in our training data.

grovduck replied:
Interesting, a bit higher loss rate than I would have guessed, but 19K *still* seems like a lot of training data.

grovduck commented on src/naip_cnn/models.py:
```diff
@@ -39,3 +39,24 @@ def CNN_v1(
     return model
+
+
+def UNet(
```
How much do you think this new model (UNet) played into the increased accuracy?  Did you try to run with `CNN_v1` as well with the 5x5 labels?  Pefhaps that model doesn't support the image labels?

aazuspan replied:
Great question! That's not something I'd tested until now because `CNN_v1` wasn't directly compatible with 2D outputs, and I also wanted to get away from the `Conv3D` layer that's typically used for video data, but after making a few modifications (changing the last `Dense` layer output shape to 25 and adding a `Reshape` layer) I was able to get it to train. 

**Using `CNN_v1` with 2016 cover data, I was able to get an MSE of 223 and an R^2 of 0.54**, which are essentially the same as we were getting with the 1D labels. It looks like the updated model might be the main difference we're seeing here, although I'll have to try training the new model with the old 1D data to confirm that - it could be a combination of the two changes. I'll run that and report back here!

grovduck replied:
Awesome summary table at the top.  Very clear how it all plays out!  UNet + 2D prediction + lots of samples for the win ...

grovduck reviewed (commented):
@aazuspan, this is incredible work.  I shouldn't be surprised by now, but I'm just blown away with what you've done here.  Other than the small authentication issue, the four notebooks worked perfectly for me and the instructions were clear.  I love the auto-filtering of data by LCMS which is a really nice feature.  And the jump in accuracy was pretty amazing as well.

My review is not completely thorough, but I found a few things I had questions about.  Just mainly excited to track your progress.

grovduck commented:
> I think that regardless we'll want to use this updated workflow with 2D label data, whether we go back to just 1x1 labels at 30m or go all the way to 150x150 labels with 1m LiDAR data

I definitely agree with this.  So much flexibility here.

aazuspan commented on src/naip_cnn/data.py:
```diff
@@ -1,160 +1,204 @@
+        """Load the full dataset."""
+        ...
+
+    def load_train(self, **kwargs):
+        return self._load(**kwargs).take(self.n_train)
+
+    def load_val(self, **kwargs):
+        return self._load(**kwargs).skip(self.n_train).take(self.n_val)
+
+    def load_test(self, **kwargs):
+        return self._load(**kwargs).skip(self.n_train + self.n_val).take(self.n_test)
```
No, this is a great point! Shuffling only happens after the split, prior to training. I just did a [quick test](https://code.earthengine.google.com/e17e50be2bad862e0fac9d56777081d5) and the points produced by `ee.Image.sample` are sorted spatially, so by doing the splits sequentially we're segregating our train, test, and validation sets. I was actually able to drop our validation MSE for canopy cover in 2016 from ~139 to ~122 by shuffling the dataset before splitting it (no improvement in R2). Whether that's actually improving accuracy by allowing our model to see more representative training data or inflating performance by validating with adjacent, autocorrelated footprints might be debatable...

Any thoughts on what the best approach is here? Shuffling is an easy solution, but we might also consider training with some datasets and validating with others, if generalization across years is a high priority. Curious to hear your thoughts! We could also pose this question to the larger group.

grovduck replied:
Actually, the question of autocorrelation crossed my mind.  The variable `MIN_SPACING` in `01_sampling.ipynb` is used in the call to `ee.Image.sample` which means that the *points* themselves are 150m apart, correct?  Does that in turn mean that the extracted footprints completely tesselate (but do not overlap) the image itself?  And how would this play in to the idea you have in #12 when image chips *would* overlap?  I obviously know next to nothing about CNNs, but is it fairly typical that the "sample" is pretty much the entire population?

> Any thoughts on what the best approach is here? Shuffling is an easy solution, but we might also consider training with some datasets and validating with others, if generalization across years is a high priority. Curious to hear your thoughts! We could also pose this question to the larger group.

Two thoughts:
1. I think that shuffling seems appropriate for the reasons you mention
2. I definitely think this is a great question for the larger group.  I know they would like to see what you've done here.

aazuspan replied:
> The variable MIN_SPACING in 01_sampling.ipynb is used in the call to ee.Image.sample which means that the points themselves are 150m apart, correct? Does that in turn mean that the extracted footprints completely tesselate (but do not overlap) the image itself? And how would this play in to the idea you have in https://github.com/lemma-osu/naip-cnn/issues/12 when image chips would overlap?

That's right, our datasets theoretically cover 100% of the LiDAR acquisitions, barring masked samples. In regards to #12, that overlap would only be used for generating predictions to reduce edge effects between chips, so it wouldn't affect model training or evaluation.

> is it fairly typical that the "sample" is pretty much the entire population?

My understanding is that it is pretty typical to use all available data, at least for LULC models, given that labeled data is usually hard to come by and CNNs are data-hungry. Since we *aren't* as data-limited, I imagine we could probably increase our spacing without any issues.

> I definitely think this is a great question for the larger group. I know they would like to see what you've done here.

Great! I figured this PR is a big enough step forward that it would be worth meeting up again to share progress and get everyone's thoughts, so I'll send out an email shortly and keep this question in mind.

For now though, it sounds like we're happy moving forward with shuffling?

grovduck replied:
> For now though, it sounds like we're happy moving forward with shuffling?

Sounds great to me.

grovduck commented:
@aazuspan, just noticed something with the output window of the predicted GeoTiffs that may or may not be an issue.  Presumably, when you're collecting the training data, I think you're using the `crsTransform` of the lidar data to set the window for the collection of the chips.  When you export the NAIP data for inference, because it doesn't look like a `crsTransform` is explicitly set, it will use the top left corner(?) to set it.  So the predicted raster will very likely have a different `crsTransform` than the lidar (this is the case with the 2016 sample data).

There is nothing at all wrong with this - the model should be equally applicable to any set of NAIP pixels - but I wonder if you want to offer the option of setting a `crsTransform` alongside the `CRS` such that predicted rasters will be snapped to either a custom window or the input lidar data.  For example, if we were using these outputs in further modeling, it would be helpful to specify the `crsTransform` to align with our other covariates.  This might be worth a separate issue or might just be a post-processing step.

aazuspan commented:
@grovduck This is a good point. I haven't put a lot of thought into the exporting process yet, and there will certainly be some big changes needed before this is ready for production. 

With the `crsTransform` issue in particular, am I understanding right that this could be fixed just by setting the `crsTransform` when starting the export in `03_export_naip.ipynb`? That should store the transform in the `mixer.json`, which would then get applied during the inference step. Do you think there's something we can add to the package to make that easier, or do we just need to modify the notebook?

aazuspan commented on scripts/convert_lidar_to_multiband.py:
I think this is a good idea, even if it should only ever need to be done once. I added a little more info to this script and committed a Code Editor script I used to process the ingested assets.

grovduck replied:
The processing script (Python version) looks great!

grovduck commented:
> With the `crsTransform` issue in particular, am I understanding right that this could be fixed just by setting the `crsTransform` when starting the export in `03_export_naip.ipynb`? That should store the transform in the `mixer.json`, which would then get applied during the inference step. Do you think there's something we can add to the package to make that easier, or do we just need to modify the notebook?

I guess my thought would be this cascade:
1. If `CRS_TRANSFORM` is defined in `config.py`, use that preferentially (this is probably what I would do to ensure that all predictions lined up with other covariates in further modeling.)
2. (This one might be tricky because the lidar dataset isn't referenced in `03_export.ipynb`).  Use the `crsTransform` of the lidar data that was used to fit the model.  This ensures the same snap to the lidar data for prediction and comparison.
3. If neither 1. or 2. is present, use the (projected) corner of the `BBOX` to set the transform.

If you can think of a helper function that could apply this logic, that would be great to add to the package, but this is definitely *not* necessary and if I want it, I can put in the notebook.  I don't want to hang up this PR on this tiny issue.

aazuspan commented:
Good thoughts on CRS transforms! In the interest of getting this merged, I opened #17 to track that separately. I'll respond to your comments there once I've spent a little more time thinking them through.

With the basic shuffling added in f5cbd08, it looks like everything here is resolved, so I'm going to go ahead and merge! I think we may want to rework the shuffling approach for cleanliness/performance in the future, but this should work for a first pass.

grovduck commented:
Looks great to me, @aazuspan.  Sorry for the delay in responding.
